lib/spack/spack/directives.py

```diff
@@ -545,7 +551,29 @@ def depends_on(spec, when=None, type=default_deptype, patches=None):
     """
 
     def _execute_depends_on(pkg):
-        _depends_on(pkg, spec, when=when, type=type, patches=patches)
+        executed = False
+        if version_translator:
+            # Produce a different dependency for every package version
+            # for which version_translator returns an answer that alters
+            # the spec.
+            for pkg_version in pkg.versions:
+                version_ish = version_translator(pkg_version)
+                if version_ish is not None:
+                    constrained_spec = spack.spec.Spec(spec)
+                    if constrained_spec.constrain(f"@{version_ish}"):
+                        pkg_depends_on(
+                            pkg,
+                            constrained_spec,
+                            when=f"@{pkg_version} {when}" if when else f"@{pkg_version}",
+                            type=type,
+                            patches=patches,
+                        )
+                        executed = True
+
+        # If we haven't already produced one or more version-specific
+        # dependencies, do the basic dependency for spec.
+        if not executed:
```

Comment:
That's not the logic of this function—if `version_translator` is defined, then a single `depends_on()` directive produces _up to_ `length(pkg.versions)` distinct dependency relationships, depending on how many non-vacuous calls to `version_translator(pkg_version)` produce a constraint that actually alters the spec. 

This number may be zero, in which case a "basic" dependency must be generated as it would be in the case where no `version_translator` is defined.